common/functions/src/scalars/dates/interval_function.rs

```diff
@@ -0,0 +1,727 @@
+
+impl IntervalFunctionFactory {
+    pub fn try_get_arithmetic_func(
+        columns: &DataColumnsWithField,
+    ) -> Option<IntervalArithmeticFunction> {
+        if columns.len() != 2 {
+            return None;
+        }
+
+        let mut interval_opt = None;
+        let mut date_datetime_opt = None;
+        columns.iter().for_each(|column| match column.data_type() {
+            DataType::Interval(_) => interval_opt = Some(column),
+            DataType::Date16 | DataType::Date32 | DataType::DateTime32(_) => {
+                date_datetime_opt = Some(column)
+            }
+            _ => {}
+        });
+
+        if interval_opt.is_none() || date_datetime_opt.is_none() {
+            return None;
+        }
+        Some(Self::get_interval_arithmetic_func(
+            interval_opt.unwrap().data_type(),
+            date_datetime_opt.unwrap().data_type(),
+        ))
+    }
+
+    //////////////////////////////////////////////////////////////////////////////////////////////
+    //  Starting from here is interval typed arithmetic functions, including:
+    //   1. interval_daytime_plus_minus_date -------- Interval(DayTime) +/- Date16/Date32
+    //   2. interval_daytime_plus_minus_datetime32 -- Interval(DayTime) +/ DateTime32
+    //   3. interval_month_plus_minus_date16 -------- Interval(YearMonth) +/ Date16
+    //   4. interval_month_plus_minus_date32 -------- Interval(YearMonth) +/ Date32
+    //   5. interval_month_plus_minus_datetime32 ---- Interval(YearMonth) +/ DateTime32
+
+    fn get_interval_arithmetic_func(
+        interval: &DataType,
+        date_datetime: &DataType,
+    ) -> IntervalArithmeticFunction {
+        match interval {
+            DataType::Interval(IntervalUnit::YearMonth) => match date_datetime {
+                DataType::Date16 => Self::interval_month_plus_minus_date16,
+                DataType::Date32 => Self::interval_month_plus_minus_date32,
+                DataType::DateTime32(_) => Self::interval_month_plus_minus_datetime32,
+                _ => unreachable!(),
+            },
+            DataType::Interval(IntervalUnit::DayTime) => match date_datetime {
+                DataType::Date32 | DataType::Date16 => Self::interval_daytime_plus_minus_date,
+                DataType::DateTime32(_) => Self::interval_daytime_plus_minus_datetime32,
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    }
+
+    fn interval_daytime_plus_minus_date(
+        op: &DataValueArithmeticOperator,
+        a: &DataColumnWithField,
+        b: &DataColumnWithField,
+    ) -> Result<DataColumn> {
+        let (interval, date) = Self::validate_input(op, a, b)?;
+        let milliseconds_per_day = 24 * 3600 * 1000;
+
+        let days: DataColumn = interval
+            .column()
+            .to_array()?
+            .i64()?
+            .apply(|ms| ms / milliseconds_per_day)
+            .into();
+        date.column().arithmetic(op.clone(), &days)
```

Comment:
If `date` is `Date16 or Date32`, the result of `date.column().arithmetic` will be  `i64` column, and it represents days.